File: id3v2/Frame.go
//go:build windows

package id3v2

import (
	"encoding/binary"
)

type Frame struct {
	name4        string
	declaredSize int // Used only at parsing.
	flags        [2]byte
	body         Body // Polymorphic.
}

func (me *Frame) Name4() string     { return me.name4 }
func (me *Frame) DeclaredSize() int { return me.declaredSize }
func (me *Frame) Body() Body        { return me.body }

// Constructor.
func _FrameParse(src []byte) (*Frame, error) {
	// Parse the 10-byte frame header.
	name4 := string(src[0:4])
	declaredSize := int(binary.BigEndian.Uint32(src[4:8]) + 10) // also count 10-byte tag header
	flags := [2]byte{src[8], src[9]}

	if declaredSize > len(src) {
		declaredSize = len(src) // if serialized with error, be complacent
	}

	src = src[10:declaredSize] // skip frame header, truncate to declared frame size

	// Parse the data body.
	body, err := _BodyParse(name4, src)
	if err != nil {
		return nil, err
	}

	return &Frame{name4, declaredSize, flags, body}, nil
}

// Constructor.
func _FrameNewWithText(name4, text string) *Frame {
	switch name4 {
	case "COMM":
		return &Frame{
			name4: name4,
			body: &BodyComment{
				Lang3: "eng",
				Descr: "",
				Text:  text,
			},
		}
	default: // otherwise assume simple text frame
		return &Frame{
			name4: name4,
			body: &BodyText{
				Text: text,
			},
		}
	}
}

// Returns a newly allocated frame with a copy of all the data.
func (me *Frame) Clone() *Frame {
	return &Frame{me.name4, me.declaredSize, me.flags, me.body.Clone()}
}

// The size in bytes required to serialize the frame.
func (me *Frame) SerializeSize() int {
	return 10 + me.body.SerializeSize() // start with 10-byte frame header
}

// Serializes the frame into bytes. The dest buffer will be appended.
func (me *Frame) Serialize(dest []byte) []byte {
	dest = append(dest, []byte(me.name4)...)

	szBody := me.body.SerializeSize() // don't count 10-byte header size
	dest = appendUint32(dest, binary.BigEndian, uint32(szBody))

	dest = append(dest, me.flags[:]...)
	dest = me.body.Serialize(dest)
	return dest
}

File: id3v2/frame.cpp
#include "frame.h"
using namespace id3v2;

Frame::Frame(std::span<const BYTE> src) {
	// Parse the 10-byte frame header.
	std::span<const BYTE> srcName4{src.subspan(0, 4)};
	_name4 = misc::parse_str_iso88591(srcName4);
	_declaredSize = misc::parse_dword_be(src.subspan(4)) + 10; // also count 10-byte tag header
	_flags = {src[8], src[9]};

	if (_declaredSize > src.size()) [[unlikely]] {
		_declaredSize = src.size(); // if serialized with error, be complacent
	}

	src = src.subspan(10, _declaredSize - 10); // skip frame header, truncate to declared frame size

	if (_name4 == L"COMM") {
		_body = std::make_unique<BodyComment>(src);
	} else if (_name4 == L"APIC") {
		_body = std::make_unique<BodyPicture>(src);
	} else if (_name4 == L"GEOB") {
		_body = std::make_unique<BodyGeob>(src);
	} else if (_name4 == L"TXXX") {
		_body = std::make_unique<BodyUserText>(src);
	} else if (_name4[0] == L'T' && _name4 != L"TDAT") [[likely]] {
		_body = std::make_unique<BodyText>(src);
	} else { // everything else is treated as raw binary
		_body = std::make_unique<BodyBinary>(src);
	}
}

Frame::Frame(wd::StrView name4, wd::StrView simpleText) {
	_name4 = name4;
	if (_name4 == L"COMM") {
		_body = std::make_unique<BodyComment>(simpleText);
	} else if (_name4[0] == L'T' && _name4 != L"TDAT") [[likely]] {
		_body = std::make_unique<BodyText>(simpleText);
	} else {
		throw ParsingError{L"Invalid body type for creating simple text."};
	}
}

Frame::Frame(const std::vector<BYTE> &picData, wd::StrView mimeType) {
	_name4 = L"APIC";
	_body = std::make_unique<BodyPicture>(picData, mimeType);
}

Frame Frame::clone() const {
	Frame f{};
	f._name4 = _name4;
	f._declaredSize = _declaredSize;
	f._flags = _flags;
	f._body = _body->clone();
	return f;
}

size_t Frame::serialize_len() const {
	return 10 + _body->serialize_len(); // start with 10-byte frame header
}

void Frame::serialize(std::vector<BYTE> &dest) const {
	misc::serialize_str(Encoding::iso88591, dest, _name4, misc::NullT::no);
	misc::serialize_dword_be(dest, static_cast<DWORD>(_body->serialize_len())); // don't count 10-byte header size
	dest.insert(dest.end(), _flags.begin(), _flags.end());
	_body->serialize(dest);
}

void Frame::set_simple_text(wd::StrView text) {
	if (auto t = dynamic_cast<BodyText*>(_body.get()); t) {
		t->text = text;
	} else if (auto u = dynamic_cast<BodyUserText*>(_body.get()); u) {
		u->descr.clear();
		u->text = text;
	} else if (auto c = dynamic_cast<BodyComment*>(_body.get()); c) {
		c->lang3 = L"eng";
		c->descr.clear();
		c->text = text;
	} else {
		throw ParsingError{L"Invalid body type for writing simple text."};
	}
}
